File: netapp-manageability-sdk-9.3/src/sample/DataFabric_Manager/API_Sample_Code/advanced/Ruby/perf_counters/perf_cpu_util.rb
#===============================================================#
#                                                               #
# ID                                                            #
#                                                               #
# perf_cpu_util.rb                                              #
#                                                               #
# Copyright (c) 2011 NetApp, Inc. All rights reserved.          #
# Specifications subject to change without notice.              #
#                                                               #
# This Sample code is supported from DataFabric Manager 3.7.1   #
# onwards.                                                      #
# However few of the functionalities of the sample code may     #
# work on older versions of DataFabric Manager.                 #
#===============================================================#

$:.unshift '../../../../../../../lib/ruby/NetApp'
require 'NaServer'


def print_help() 
	print ("\nCommand:\n")
	print ("perf_cpu_util.rb <dfm> <user> <password> <storage-system>\n")
	print ("<dfm>            -- DFM Server name\n")
	print ("<user>           -- User name\n")
	print ("<password>       -- Password\n")
	print ("<storage-system> -- Storage system\n")
	print ("-" * 80 + "\n")
	print ("This sample code prints CPU utilization statistics of a storage \n")
	print ("system. The sample code collects CPU utilization data for 2 weeks\n")
	print ("and prints the data in a format, which enables comparision of CPU\n")
	print ("utilization in day, hour format for both the weeks\n")
	print ("Output data of this sample code can be used to generate chart.\n")
	print ("To generate the graph, redirect output of this sample code to\n")
	print ("an Excel sheet.\n")
	exit 
end


def per_week_data(server_ctx, start_time, end_time)
	perf_in = NaElement.new("perf-get-counter-data")
	perf_in.child_add_string("start-time", start_time.asctime)
	perf_in.child_add_string("end-time", end_time.asctime)
	perf_in.child_add_string("sample-rate", 3600)
	perf_in.child_add_string("time-consolidation-method", "average")
	instance_info = NaElement.new("instance-counter-info")
	instance_info.child_add_string("object-name-or-id", ARGV[3])
	counter_info = NaElement.new("counter-info")
	perf_obj_ctr = NaElement.new("perf-object-counter")
	perf_obj_ctr.child_add_string("object-type", "system")
	perf_obj_ctr.child_add_string("counter-name", "avg_processor_busy")
	counter_info.child_add(perf_obj_ctr)
	instance_info.child_add(counter_info)
	perf_in.child_add(instance_info)
	perf_out = server_ctx.invoke_elem(perf_in)	
	if(perf_out.results_status() == "failed") 
		print(perf_out.results_reason() +"\n")
		exit
	end	
	return perf_out
end


def get_time_arr(perf_out) 
	i = 0
	time_arr = []
	instance = perf_out.child_get("perf-instances")
	instances = instance.children_get()
	instances.each do |rec|
		counters = rec.child_get("counters")
		perf_cnt_data = counters.children_get()		
		perf_cnt_data.each do |rec1|
			counter_str = rec1.child_get_string("counter-data")
			counter_arr = counter_str.split(',')			
			counter_arr.each do |time_val|
				time_val_arr = time_val.split(':')
				time_arr[i] = time_val_arr[0].to_f
				i = i + 1
			end
		end
	end	
	return time_arr
end


def get_data_arr(perf_out)
	i = 0
	data_arr = []
	instance = perf_out.child_get("perf-instances")
	instances = instance.children_get()
	instances.each do |rec|
		counters = rec.child_get("counters")
		perf_cnt_data = counters.children_get()		
		perf_cnt_data.each do |rec1|
			counter_str = rec1.child_get_string("counter-data")
			counter_arr = counter_str.split(',')			 
			counter_arr.each do |time_val|
				time_val_arr = time_val.split(':')
				data_arr[i] = time_val_arr[1].to_f
				i = i + 1
			end
		end
	end	
	return data_arr
end


def print_output() 
	print ("Week1\t\t\t\t\t\t\t\t\t\tWeek2\t\n")
	print ("Time\t\t\t\tCPU Busy\t\t\t\t\tTime\t\t\t\tCPU Busy\n")
	i = 0
	j = 0	
	while(i < $time_arr1.length and j < $time_arr2.length)		
		if (($time_arr2[j] - $time_arr1[i]) > 608400) 
			print (Time.at($time_arr1[i]).to_s + "\t" + $data_arr1[i].to_s + "\t\t")
			i = i + 1
			print ("\t\t\t")		
		elsif (($time_arr2[j] - $time_arr1[i]) < 601200)
			print (Time.at($time_arr2[j]))
			print ("\t" + $data_arr2[j].to_s + "\t\t\t")
			j = j + 1			
		else 
			print(Time.at($time_arr1[i]).to_s + "\t" + $data_arr1[i].to_s + "\t\t")
			i = i + 1
			print (Time.at($time_arr2[j]).to_s + "\t" + $data_arr2[j].to_s + "\t\t\t")
			j = j + 1
		end
		print("\t\t\t ")
	end	
	print("")	
	while(i < $time_arr1.length) 
        print( Time.at($time_arr1[i]).to_s + "\t" + $data_arr1[i].to_s + " \t\t\t\t\t\t")
		i = i + 1
		print(" ")
	end    
    while(j < $time_arr2.length)
        print (Time.at($time_arr2[j]).to_s + "\t" + $data_arr2[j].to_s + "\t\t\t\t\t\t ")
		j = j + 1
		print(" ")
	end
end

    
def main() 	
	dfm = ARGV[0]
	user = ARGV[1]
	pw  = ARGV[2]
	
	# Initialize server context
	server_ctx = NaServer.new(dfm, 1, 0)
	server_ctx.set_admin_user(user, pw)
	server_ctx.set_server_type("DFM")

	# Start time and end time for week1 data collection
	#Start time = current time - (60sec * 60min * 24hrs * 14days)
	starttime1 = Time.now - 1209600
	#end time = current time - (60sec * 60min * 24hrs * 7days)
	endtime1 = Time.now - 604800

	# Start time and end time for week2 data collection
	#Start time = current time - (60sec * 60min * 24hrs * 7days)
	starttime2 = Time.now - 604800
	# end time = current time
	endtime2 = Time.now

	# Collect data for Week1
	perf_out1 = per_week_data(server_ctx, starttime1, endtime1)

	# Collect data for Week2
	perf_out2 = per_week_data(server_ctx, starttime2, endtime2)
	$time_arr1 = get_time_arr(perf_out1)
	$time_arr2 = get_time_arr(perf_out2)
	$data_arr1 = get_data_arr(perf_out1)
	$data_arr2 = get_data_arr(perf_out2)
	print_output()
	exit
end

	
# Command line arguments
args = ARGV.length
# check for valid number of parameters
if (args != 4)
	print_help()
end
$time_arr1 = []
$time_arr2 = []
$data_arr1 = []
$data_arr2 = []
#Invoke routine
main()

#=========================== POD ============================#

'=head1 NAME

  perf_cpu_util.rb - Prints CPU utilization statistics. This sample code
  compares CPU utilization statistics of 2 weeks.


=head1 SYNOPSIS

  perf_cpu_util.rb <dfm> <user> <password> <storage-system>

=head1 ARGUMENTS

  <dfm>
  DFM Server name.

  <user>
  User name.

  <password>
  Password.

  <storage-system>
  Storage system.

=head1 SEE ALSO

  NaElement.rb, NaServer.rb

=head1 COPYRIGHT

 Copyright (c) 2009 NetApp, Inc. All rights reserved.
 Specifications subject to change without notice.

=cut'

File: netapp-manageability-sdk-9.3/src/sample/DataFabric_Manager/API_Sample_Code/advanced/Java/perf_counters/perfCpuUtil.java
/**
 * $Id:$
 *
 * perfCpuUtil.java
 *
 * Copyright (c) 2009 NetApp, Inc. All rights reserved.
 * Specifications subject to change without notice.
 *
 *
 * This Sample code is supported from DataFabric Manager 3.7.1
 * onwards.
 * However few of the functionalities of the sample code may
 * work on older versions of DataFabric Manager.
 */

import netapp.manage.NaElement;
import netapp.manage.NaServer;
import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;
import java.util.List;

public class perfCpuUtil {
    private static NaServer server;
    private static String[] Arg;
    private static ArrayList timeArray = new ArrayList();
    private static ArrayList dataArray = new ArrayList();

    public static void print_help() {
        /*
         * Print usage information for cpu util.
         */
        System.out.println("\nCommand:");
        System.out.println("perf_cpu_util.pl <dfm> <user> <password>"
                + " <storage-system>");
        System.out.println("<dfm>            -- DFM Server name");
        System.out.println("<user>           -- User name");
        System.out.println("<password>       -- Password");
        System.out.println("<storage-system> -- Storage system");
        System.out.println("-----------------------------------------"
                + "---------------------------\n");
        System.out.println("This sample code prints CPU utilization"
                + " statistics of a storage ");
        System.out.println("system. The sample code collects CPU "
                + "utilization data for 2 weeks");
        System.out.println("and prints the data in a format, which enables"
                + " comparision of CPU");
        System.out.println("utilization in day, hour format for both the"
                + " weeks");
        System.out.println("Output data of this sample code can be used "
                + "to generate chart.");
        System.out.println("To generate the graph, redirect output of this"
                + " sample code to");
        System.out.println("an Excel sheet.");

        System.exit(1);
    }

    public static void main(String[] args) {

        Arg = args;
        int arglen = Arg.length;
        long startTime1 = (new Date().getTime() / 1000) - 1209600;
        long endTime1 = (new Date().getTime() / 1000) - 604800;
        long startTime2 = (new Date().getTime() / 1000) - 604800;
        long endTime2 = (new Date().getTime() / 1000);

        // Checking for valid number of parameters
        if (arglen != 4)
            print_help();

        String dfmserver = Arg[0];
        String dfmuser = Arg[1];
        String dfmpw = Arg[2];

        try {
            // Initialize connection to server, and
            // request version 1.0 of the API set
            //
            // Creating a server object and setting appropriate attributes
            server = new NaServer(dfmserver, 1, 0);
            server.setStyle(NaServer.STYLE_LOGIN_PASSWORD);
            server.setServerType(NaServer.SERVER_TYPE_DFM);

            server.setAdminUser(dfmuser, dfmpw);

            // Collect data for Week1
            NaElement perfOut1 = perWeekData(Long.toString(startTime1),
                    Long.toString(endTime1));

            // Collect data for Week2
            NaElement perfOut2 = perWeekData(Long.toString(startTime2),
                    Long.toString(endTime2));

            // Extracting the time and date
            getTimedataArrayay(perfOut1);
            ArrayList timeArray1 = new ArrayList(timeArray.size());
            ArrayList dataArray1 = new ArrayList(dataArray.size());
            timeArray1.addAll(timeArray);
            dataArray1.addAll(dataArray);
            // clearing arraylist as they are global arraylists
            timeArray.clear();
            dataArray.clear();

            // values will be loaded into global arraylists
            getTimedataArrayay(perfOut2);
            System.out.println("Week1\t\tWeek2\t");
            System.out.println("Time\tCPU Busy\tTime\tCPU Busy");
            int i;
            int j;

            for (i = 0, j = 0; i < timeArray1.size() && j < timeArray.size();) {
                if ((Long.parseLong(timeArray.get(j).toString()) - Long
                        .parseLong(timeArray1.get(i).toString())) > 608400) {
                    System.out.print(new Date(Long.parseLong(timeArray1.get(i)
                            .toString()) * 1000)
                            + "\t"
                            + dataArray1.get(i).toString() + "\t");
                    i++;
                    System.out.println("\t\t");
                } else if ((Long.parseLong(timeArray.get(j).toString()) - Long
                        .parseLong(timeArray1.get(i).toString())) < 601200) {
                    System.out.print("\t\t");
                    System.out.println(new Date(Long.parseLong(timeArray.get(j)
                            .toString()) * 1000)
                            + "\t"
                            + dataArray.get(j).toString());
                    j++;
                } else {
                    System.out.print(new Date(Long.parseLong(timeArray1.get(i)
                            .toString()) * 1000)
                            + "\t"
                            + dataArray1.get(i).toString() + "\t");
                    i++;
                    System.out.println(new Date(Long.parseLong(timeArray.get(j)
                            .toString()) * 1000)
                            + "\t"
                            + dataArray.get(j).toString());
                    j++;
                }
            }
            while (i < timeArray1.size()) {
                System.out.print(new Date(Long.parseLong(timeArray1.get(i)
                        .toString()) * 1000)
                        + "\t"
                        + dataArray1.get(i).toString() + "\t");
                i++;
                System.out.println("\t\t");
            }
            while (j < timeArray.size()) {
                System.out.print("\t\t");
                System.out.println(new Date(Long.parseLong(timeArray.get(j)
                        .toString()) * 1000)
                        + "\t"
                        + dataArray.get(j).toString());
                j++;
            }
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }

    }

    public static NaElement perWeekData(String startTime, String endTime) {
        NaElement perfOut = null;
        try {

            String storageSystem = Arg[3];

            NaElement perfIn = new NaElement("perf-get-counter-data");

            perfIn.addNewChild("start-time", startTime);
            perfIn.addNewChild("end-time", endTime);
            perfIn.addNewChild("sample-rate", "3600");
            perfIn.addNewChild("time-consolidation-method", "average");

            NaElement instanceInfo = new NaElement("instance-counter-info");
            instanceInfo.addNewChild("object-name-or-id", storageSystem);

            NaElement counterInfo = new NaElement("counter-info");
            NaElement perfObjCtr = new NaElement("perf-object-counter");
            perfObjCtr.addNewChild("object-type", "system");
            perfObjCtr.addNewChild("counter-name", "avg_processor_busy");

            counterInfo.addChildElem(perfObjCtr);
            instanceInfo.addChildElem(counterInfo);
            perfIn.addChildElem(instanceInfo);

            perfOut = server.invokeElem(perfIn);

        } catch (Exception e) {
            System.err.println(e.toString());
            System.exit(1);
        }
        return perfOut;
    }

    public static void getTimedataArrayay(NaElement perfOut) {

        String[] counterArray;
        String[] timeValArray;
        int read;

        NaElement instance = perfOut.getChildByName("perf-instances");

        List instances = null;
        if (instance != null)
            instances = instance.getChildren();
        if (instances == null)
            return;

        Iterator instanceIter = instances.iterator();

        while (instanceIter.hasNext()) {
            NaElement rec = (NaElement) instanceIter.next();
            NaElement counters = rec.getChildByName("counters");

            List perfCntData = null;
            if (counters != null)
                perfCntData = counters.getChildren();
            if (perfCntData == null)
                return;

            Iterator counterIter = perfCntData.iterator();

            while (counterIter.hasNext()) {
                NaElement rec1 = (NaElement) counterIter.next();
                String counterStr = rec1.getChildContent("counter-data");
                counterArray = counterStr.split(",");

                for (read = 0; read < counterArray.length; read++) {
                    timeValArray = counterArray[read].split(":");
                    timeArray.add(timeValArray[0]);
                    dataArray.add(timeValArray[1]);
                }
            }
        }
    }
}